src/VisualStudio/Core/Impl/CodeModel/ProjectCodeModelFactory.cs

```diff
@@ -66,61 +65,65 @@ public ProjectCodeModelFactory(
-        ImmutableSegmentedList<DocumentId> documentIds, CancellationToken cancellationToken)
+    private async ValueTask ProcessNextWorkspaceChangeEventBatchAsync(
+        ImmutableSegmentedList<WorkspaceChangeEventArgs> workspaceChangeEvents, CancellationToken cancellationToken)
     {
-        // This logic preserves the previous behavior we had with IForegroundNotificationService.
-        // Specifically, we don't run on the UI thread for more than 15ms at a time.  And once we 
-        // have, we wait 50ms before continuing.  These constants are just what we defined from
-        // legacy, and otherwise have no special meaning.
-        const int MaxTimeSlice = 15;
-        var delayBetweenProcessing = DelayTimeSpan.NearImmediate;
-
         Debug.Assert(!_threadingContext.JoinableTaskContext.IsOnMainThread, "The following context switch is not expected to cause runtime overhead.");
         await TaskScheduler.Default;
 
-        // Ensure MEF services used by the code model are initially obtained on a background thread.
-        // This code avoids allocations where possible.
-        // https://github.com/dotnet/roslyn/issues/54159
-        string? previousLanguage = null;
-        foreach (var projectState in _visualStudioWorkspace.CurrentSolution.SolutionState.SortedProjectStates)
-        {
-            if (projectState.Language == previousLanguage)
-            {
-                // Avoid duplicate calls if the language did not change
-                continue;
-            }
+        // Calculate the full set of changes over this set of events while on a background thread.
+        using var _1 = PooledHashSet<DocumentId>.GetInstance(out var documentIds);
+        AddChangedDocuments(workspaceChangeEvents, documentIds);
 
-            previousLanguage = projectState.Language;
-            projectState.LanguageServices.GetService<ICodeModelService>();
-            projectState.LanguageServices.GetService<ISyntaxFactsService>();
-            projectState.LanguageServices.GetService<ICodeGenerationService>();
-        }
+        if (documentIds.Count == 0)
+            return;
+
+        // get the file path information while on a background thread
+        using var _2 = ArrayBuilder<(ProjectCodeModel projectCodeModel, string filename)>.GetInstance(out var projectCodeModelAndFileNames);
+        AddProjectCodeModelAndFileNames(documentIds, projectCodeModelAndFileNames);
+
+        // Ensure MEF services used by the code model are initially obtained on a background thread.
+        EnsureServicesCreated();
 
         await _threadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
 
-        var stopwatch = SharedStopwatch.StartNew();
-        foreach (var documentId in documentIds)
+        // Fire off the code model events while on the main thread
+        await FireEventsForChangedDocumentsAsync(projectCodeModelAndFileNames, cancellationToken).ConfigureAwait(false);
```

Comment:
nit: since this method is async, you can move the STMTA into it.

Reply:
I actually kind of preferred it in the outer method, as that's what's done for the background thread assurance. This way, if more work needs to be done on the main thread, it can just be called from the same section in ProcessNextWorkspaceChangeEventBatchAsync.